### ISSUE TITLE
nn.binaryCrossEntropy NaNs on saturated (converged) logits -- 0 * log(0); needs the BCEWithLogits formulation

### ISSUE THREAD

**Issue by johnhenry** [johnhenry/math-plus#85]
Found while building mallory-graph's ML playground (its issue #34): `nn.binaryCrossEntropy` returns **NaN once the classifier converges well**, poisoning the weights on the next `backward()`.

## Repro (deterministic)

```ts
import { nn, variable } from "mallory-tensor-autograd";
import { Tensor } from "mallory-tensor-core";

const z = variable(Tensor.from([50, -50], { dtype: "f64" }).reshape([2, 1])); // confident, CORRECT logits
const y = variable(Tensor.from([1, 0], { dtype: "f64" }).reshape([2, 1]));
nn.binaryCrossEntropy(z, y).value.item(); // NaN (expected ~0)
```

In a real training run: a 2->8->1 MLP on a two-moons set (data seed 7, model seed 42, Adam lr 0.05) NaNs at epoch ~249 -- i.e. ~49 epochs after the loss reached ~1e-2. The better the fit, the sooner it dies.

## Root cause

The current implementation computes probabilities first, then logs them:

```ts
const p = logits.sigmoid();
const term1 = target.mul(p.log());
const term2 = target.mul(-1).add(1).mul(p.mul(-1).add(1).log());
```

For |z| ≳ 37, f64 `sigmoid(z)` saturates to exactly `1.0` (or `0.0`), so the *inactive* side's factor becomes `0 * log(0) = 0 * -Inf = NaN`. The multiplication by a zero target does not rescue it -- IEEE 754 `0 * Inf` is NaN by definition. Note the sibling `crossEntropy`'s own doc comment already flags this class of concern ("softmax ... is safely > 0 for the logit magnitudes a toy training loop produces") -- binary classification breaks that assumption in practice, because a *converged* toy classifier routinely exceeds |z|=37.

## Fix

Compute from logits, the standard `BCEWithLogits` formulation:

```
L(z, y) = relu(z) - z*y + log(1 + exp(-|z|))
```

If adding `exp`/`abs` Variable ops is out of scope, the last term is expressible with existing ops as `-log(sigmoid(|z|))` with `|z| = relu(z) + relu(-z)` -- `sigmoid(|z|)` is always >= 0.5 so its log never sees 0. mallory-graph now carries exactly this as an app-side workaround (`stableBinaryCrossEntropy` in its `src/lib/ml-playground.ts`), verified equal to the current implementation to ~1e-12 in the non-saturated regime; it also carries a regression test that asserts the upstream NaN still exists, so the workaround gets retired when this is fixed.

Consider the same audit for anything else that logs a saturating quantity (`p.log()` patterns generally).
